src/webgpu/util/math.ts

```diff
@@ -149,29 +118,98 @@ export function nextAfter(val: number, dir: boolean = true, flush: boolean): Sca
   const converted: number = new Float32Array([val])[0];
-  assert(val === converted, `${val} is not expressible as a f32.`);
-
-  const u32_val = new Uint32Array(new Float32Array([val]).buffer)[0];
-  const is_positive = (u32_val & 0x80000000) === 0;
-  let result = u32_val;
-  if (dir === is_positive) {
-    result += 1;
+  let u32_result: number;
+  if (val === converted) {
+    // val is expressible precisely as a float32
+    let u32_val: number = new Uint32Array(new Float32Array([val]).buffer)[0];
+    const is_positive = (u32_val & 0x80000000) === 0;
+    if (dir === is_positive) {
+      u32_val += 1;
+    } else {
+      u32_val -= 1;
+    }
+    u32_result = flush ? flushSubnormalBits(u32_val) : u32_val;
   } else {
-    result -= 1;
+    // val had to be rounded to be expressed as a float32
+    if (dir === converted > val) {
+      // Rounding was in the direction requested
+      u32_result = new Uint32Array(new Float32Array([converted]).buffer)[0];
+    } else {
+      // Round was opposite of the direction requested, so need nextAfter in the requested direction.
+      // This will not recurse since converted is guaranteed to be a float32 due to the conversion above.
+      const next = nextAfter(converted, dir, flush).value.valueOf() as number;
+      u32_result = new Uint32Array(new Float32Array([next]).buffer)[0];
+    }
   }
-  result = flush ? flushSubnormalBits(result) : result;
 
   // Checking for overflow
-  if ((result & 0x7f800000) === 0x7f800000) {
+  if ((u32_result & 0x7f800000) === 0x7f800000) {
     if (dir) {
       return f32Bits(kBit.f32.infinity.positive);
     } else {
       return f32Bits(kBit.f32.infinity.negative);
     }
   }
-  return f32Bits(result);
+
+  return f32Bits(u32_result);
+}
+
+/**
+ * @returns ulp(x), the unit of least precision for a specific number as a 32-bit float
+ *
+ * ulp(x) is the distance between the two floating point numbers nearest x.
+ * This value is also called unit of last place, ULP, and 1 ULP.
+ * See the WGSL spec and http://www.ens-lyon.fr/LIP/Pub/Rapports/RR/RR2005/RR2005-09.pdf
+ * for a more detailed/nuanced discussion of the definition of ulp(x).
+ *
+ * @param target number to calculate ULP for
+ * @param flush should subnormals be flushed to zero
+ */
+export function oneULP(target: number, flush: boolean): number {
+  if (Number.isNaN(target)) {
+    return Number.NaN;
+  }
+
+  // For values at the edge of the range or beyond ulp(x) is defined as
+  // the distance between the two nearest representable numbers to the
+  // appropriate edge.
+  if (target === Number.POSITIVE_INFINITY || target >= kValue.f32.positive.max) {
+    return kValue.f32.positive.max - kValue.f32.positive.nearest_max;
+  } else if (target === Number.NEGATIVE_INFINITY || target <= kValue.f32.negative.min) {
+    return kValue.f32.negative.nearest_min - kValue.f32.negative.min;
+  } else {
+    const b = nextAfter(target, true, flush).value.valueOf() as number;
```

Comment:
As discussed, I'm not sure about this - it seems like the ULP is (nearly always) twice what it should be.

Reply:
Done